### PR TITLE
power: Eliminate SYS_PM_* power states.

### PR DESCRIPTION
**Pull request by pizi-nordic**
The power management framework used two different abstractions
to describe power states. The SYS_PM_* given coarse information
what kind of power state (low power or deep sleep) was used,
while the SYS_POWER_STATE_* abstraction provided information
about particular power mode.

This PR removes the SYS_PM_* abstraction as the same
information is already carried in SYS_POWER_STATE_*.

### CODE REVIEW

**Review by pdunaj — APPROVED** [review]
seems ok but we will have to update our app...

**Review by dbkinder — APPROVED** [review]
doc change looks OK

**Review by anangl — APPROVED** [review]

File: soc/arc/quark_se_c1000_ss/power.c [hunk 1]
```diff
@@ -72,8 +72,10 @@ void sys_power_state_post_ops(enum power_states state)
 #if (defined(CONFIG_SYS_POWER_LOW_POWER_STATE))
 	case SYS_POWER_STATE_CPU_LPS_1:
 		/* Expire the timer as it is disabled in SS2. */
-		u32_t limit = _arc_v2_aux_reg_read(_ARC_V2_TMR0_LIMIT);
-		_arc_v2_aux_reg_write(_ARC_V2_TMR0_COUNT, limit - 1);
+		{
+			u32_t limit = _arc_v2_aux_reg_read(_ARC_V2_TMR0_LIMIT);
+			_arc_v2_aux_reg_write(_ARC_V2_TMR0_COUNT, limit - 1);
+		}
```
Comment (mnkp) [inline soc/arc/quark_se_c1000_ss/power.c]: Seems like this change is not related. Added by accident?

File: subsys/power/policy/policy_dummy.c [hunk 2]
```diff
@@ -9,71 +9,33 @@
...
-#endif
-};
-
-int sys_pm_policy_next_state(s32_t ticks, enum power_states *pm_state)
+enum power_states sys_pm_policy_next_state(s32_t ticks)
 {
-	static int cur_pm_idx;
-	int i = cur_pm_idx;
+	static u8_t cur_power_state;
+	u8_t i = cur_power_state;
```
Comment (mnkp) [inline subsys/power/policy/policy_dummy.c]: Why `u8_t`? Shouldn't it be `enum power_states`

File: subsys/power/policy/policy_residency.c [hunk 3]
```diff
@@ -15,78 +15,57 @@ LOG_MODULE_DECLARE(power);
...
-	{SYS_POWER_STATE_DEEP_SLEEP_2, SYS_PM_DEEP_SLEEP,
-			CONFIG_PM_DEEP_SLEEP_2_MIN_RES * SECS_TO_TICKS},
+	CONFIG_PM_DEEP_SLEEP_2_MIN_RES * SECS_TO_TICKS,
 #endif
 };
 
-int sys_pm_policy_next_state(s32_t ticks, enum power_states *pm_state)
+enum power_states sys_pm_policy_next_state(s32_t ticks)
 {
 	int i;
```
Comment (mnkp) [inline subsys/power/policy/policy_residency.c]: It's probably better to change `int` to `enum power_states`.

File: soc/arc/quark_se_c1000_ss/power.c (hunk 1 above)
Comment (pizi-nordic) [reply soc/arc/quark_se_c1000_ss/power.c]: No, this fixes an compilation issue introduced in one of my previous PRs and not detected by CI.
I hoped for fast merge :)

File: subsys/power/policy/policy_dummy.c (hunk 2 above)
Comment (pizi-nordic) [reply subsys/power/policy/policy_dummy.c]: This is `u8_t` as we are doing mathematics operations on this variable later.

File: subsys/power/policy/policy_residency.c (hunk 3 above)
Comment (pizi-nordic) [reply subsys/power/policy/policy_residency.c]: No. The `int i` is used as an index to array.

File: soc/arc/quark_se_c1000_ss/power.c (hunk 1 above)
Comment (mnkp) [reply soc/arc/quark_se_c1000_ss/power.c]: I don't mind sneaking it into this PR but shoudn't it be a separate commit?

File: subsys/power/policy/policy_dummy.c (hunk 2 above)
Comment (mnkp) [reply subsys/power/policy/policy_dummy.c]: In C arithmetic operations on enum are legal (but please correct me if I'm wrong). We also want to conform to Misra-C and we have this
```
	if (!sys_pm_ctrl_is_state_enabled(i)) {
```
with no cast. I am not sure any more what's the best way to handle this. In any case enum maps to `int`, not `u8_t`.

@ceolin could you have a quick look if this code conforms to Misra-C?

Comment (pizi-nordic) [reply subsys/power/policy/policy_dummy.c]: Arithmetic on enum is possible in C as long as it stays in bounds of the enum. If we cross the boundary, the behaviour is unspecified. In C++ direct arithmetic on enum is not allowed.
I personally would like to keep the integers. If MISRA requires explicit casts, I can add them.

Comment (ceolin) [reply subsys/power/policy/policy_dummy.c]: this is not allowed, enums are a specific "essential type" in MISRA-C 2012 terminology and promote int (signed, unsigned, ...) are not allowed. We've to cast `i` here.

Comment (ceolin) [reply subsys/power/policy/policy_dummy.c]: this is not allowed, enums are a specific "essential type" in MISRA-C 2012 terminology and promote int (signed, unsigned, ...) are not allowed. We've to cast `i` here.

Comment (mnkp) [reply subsys/power/policy/policy_dummy.c]: Misra-C requires more careful type handling, explicit casting, etc. But I don't remember any more rules on enum to integer type conversion. We can fix it later, if required. Still, I believe we should change `u8_t` to `int`. Enums are stored as `int` (unless compiler option is forcing compact/short enums).

Comment (ceolin) [reply subsys/power/policy/policy_dummy.c]: > Misra-C requires more careful type handling, explicit casting, etc. But I don't remember any more rules on enum to integer type conversion. We can fix it later, if required. Still, I believe we should change `u8_t` to `int`. Enums are stored as `int` (unless compiler option is forcing compact/short enums).

this violates rule 10.3, a simple cast is enough

Comment (mnkp) [reply subsys/power/policy/policy_dummy.c]: @ceolin thanks for clarification!

File: soc/arc/quark_se_c1000_ss/power.c (hunk 1 above)
Comment (pizi-nordic) [reply soc/arc/quark_se_c1000_ss/power.c]: Change removed.

File: subsys/power/policy/policy_dummy.c (hunk 2 above)
Comment (pizi-nordic) [reply subsys/power/policy/policy_dummy.c]: Added casts, changed type to int. I have left `static u8_t cur_power_state` to avoid wasting 3 bytes of ram.

**Review by mnkp — APPROVED** [review]

File: doc/reference/power_management/index.rst [hunk 4]
```diff
@@ -184,26 +172,26 @@ The power management subsystem classifies power management schemes
...
 
 CPU is power gated and loses execution context. Execution will resume at
 OS startup code or at a resume point determined by a bootloader that supports
 deep sleep resume. Depending on the SOC's implementation of the power saving
 feature, it may turn off power to most devices. RAM may be retained by some
 implementations, while others may remove power from RAM saving considerable
-power. Power states in this category save more power than
-`SYS_PM_LOW_POWER_STATE`_ and would have higher wake latencies.
+power. Power states in this category save more power than Low Power states
+and would have higher wake latencies.
```
Comment (ioannisg) [inline doc/reference/power_management/index.rst]: wake up?

Comment (pizi-nordic) [reply doc/reference/power_management/index.rst]: In whole document there is "wake latency" repeated multiple times. Should I change it here or everywhere?

Comment (ioannisg) [reply doc/reference/power_management/index.rst]: I would change this everywhere - "wake latency" does not make too much sense to me. I spotted it here, because this text was changed :) "wake-up latency" is the right term, IMO

**Review by nashif — APPROVED** [review]
